controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -9,515 +9,405 @@
-            return FutureHelpers.failedFuture(new DataNotFoundException(stream));
-        }
-
-        @Override
-        public CompletableFuture<Void> setColdMarker(final int segmentNumber, final long timestamp) {
-            return FutureHelpers.failedFuture(new DataNotFoundException(stream));
-        }
-
-        @Override
-        public CompletableFuture<Long> getColdMarker(final int segmentNumber) {
-            return FutureHelpers.failedFuture(new DataNotFoundException(stream));
-        }
-
-        @Override
-        public CompletableFuture<Void> removeColdMarker(final int segmentNumber) {
-            return FutureHelpers.failedFuture(new DataNotFoundException(stream));
-        }
-
-        @Override
-        public CompletableFuture<VersionedTransactionData> createTransaction(final UUID txId,
-                                                                             final long lease,
-                                                                             final long maxExecutionTime,
-                                                                             final long scaleGracePeriod) {
-            return FutureHelpers.failedFuture(new DataNotFoundException(stream));
-        }
-
-        @Override
-        public CompletableFuture<VersionedTransactionData> pingTransaction(final UUID txId,
-                                                                           final long lease) {
-            return FutureHelpers.failedFuture(new DataNotFoundException(stream));
-        }
-
-        @Override
-        public CompletableFuture<VersionedTransactionData> getTransactionData(final UUID txId) {
-            return FutureHelpers.failedFuture(new DataNotFoundException(stream));
-        }
-
-        @Override
-        public CompletableFuture<SimpleEntry<TxnStatus, Integer>> sealTransaction(final UUID txId,
-                                                                                  final boolean commit,
-                                                                                  final Optional<Integer> version) {
-            return FutureHelpers.failedFuture(new DataNotFoundException(stream));
-        }
+    private long creationTime = Long.MIN_VALUE;
+    private StreamConfiguration configuration;
+    private final AtomicReference<Data<Integer>> state = new AtomicReference<>();
+    private final AtomicReference<Data<Integer>> segmentTable = new AtomicReference<>();
+    private final AtomicReference<Data<Integer>> historyTable = new AtomicReference<>();
+    private final AtomicReference<Data<Integer>> indexTable = new AtomicReference<>();
+    private final ConcurrentHashMap<String, Data<Integer>> activeTxns = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, Data<Integer>> completedTxns = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<Integer, Data<Integer>> markers = new ConcurrentHashMap<>();
+    private final Object epochLock = new Object();
```

Comment:
It makes sure that we delete the epoch entry if and only if there are no transactions against that epoch.